From 748c1f30a6a202db44115f451487690046835211 Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Fri, 15 Aug 2025 19:43:56 +0100
Subject: [PATCH] baseprovider: add MD5 and SHA1

Open up access to MD5 and SHA1 from the base provider. This allows
access to these algorithms for non-security purposes from higher level
languages such as dotnet, python, ruby and others.

Crutially, this does not allow fips provider to use them for
cryptographically secure purposes such as calculating HMAC, DRBG,
PRNG, Signatures, KDF, etc.
---
 providers/baseprov.c   | 14 ++++++++++++++
 test/fips-and-base.cnf |  4 ++--
 2 files changed, 16 insertions(+), 2 deletions(-)

diff --git a/providers/baseprov.c b/providers/baseprov.c
index ad08765f00..cdb5937abb 100644
--- a/providers/baseprov.c
+++ b/providers/baseprov.c
@@ -64,6 +64,18 @@ static int base_get_params(void *provctx, OSSL_PARAM params[])
     return 1;
 }
 
+/*
+ * open up access to MD5 and SHA1 for digest/crc non-security
+ * purposes, does not allow to be used in HMAC/DRBG/PRNG/KEM/Signature
+ * by the fips provider, this allows us to completely remove SHA1 from
+ * the fips provider.
+ */
+static const OSSL_ALGORITHM base_digests[] = {
+    { PROV_NAMES_SHA1, "provider=base", ossl_sha1_functions },
+    { PROV_NAMES_MD5, "provider=base", ossl_md5_functions },
+    { NULL, NULL, NULL }
+};
+
 static const OSSL_ALGORITHM base_encoder[] = {
 #define ENCODER_PROVIDER "base"
 #include "encoders.inc"
@@ -104,6 +116,8 @@ static const OSSL_ALGORITHM *base_query(void *provctx, int operation_id,
         return base_encoder;
     case OSSL_OP_DECODER:
         return base_decoder;
+    case OSSL_OP_DIGEST:
+        return base_digests;
     case OSSL_OP_STORE:
         return base_store;
     case OSSL_OP_RAND:
diff --git a/test/fips-and-base.cnf b/test/fips-and-base.cnf
index f233f83062..9fe4a828e4 100644
--- a/test/fips-and-base.cnf
+++ b/test/fips-and-base.cnf
@@ -9,10 +9,10 @@ config_diagnostics = 1
 providers = provider_sect
 # You MUST uncomment the following line to operate in a FIPS approved manner,
 # It is commented out here purely for testing purposes.
-#alg_section = evp_properties
+alg_section = evp_properties
 
 [evp_properties]
-default_properties = "fips=yes"
+default_properties = "?fips=yes"
 
 [provider_sect]
 fips = fips_sect
-- 
2.53.0

